### PR TITLE
[x] switch to new versions of guppy and camino

### PR DESCRIPTION
**Pull request by sunshowers**
`camino` brings about `Utf8Path` and `Utf8PathBuf` wrappers, which
together simplify our path handling code quite a bit.

Going to release new versions of guppy and friends before landing this.

### CODE REVIEW

File: devtools/x/src/lint/allowed_paths.rs [hunk 1]
```diff
@@ -32,14 +32,7 @@ impl FilePathLinter for AllowedPaths {
...
         out: &mut LintFormatter<'l, '_>,
     ) -> Result<RunStatus<'l>> {
-        let file_path = match ctx.file_path().to_str() {
-            Some(file_path) => file_path,
-            None => {
-                out.write(LintLevel::Error, "path isn't valid Unicode");
-                return Ok(RunStatus::Executed);
-            }
-        };
-        if !self.allowed_regex.is_match(file_path) {
```
Comment (sunshowers) [inline devtools/x/src/lint/allowed_paths.rs:42]: Here's an example of code that gets simplified.

File: devtools/x-core/src/git.rs [hunk 2]
```diff
@@ -2,7 +2,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use crate::errors::*;
-use determinator::Paths0;
+use determinator::Utf8Paths0;
```
Comment (bmwill) [inline devtools/x-core/src/git.rs:5]: why is Utf8Paths0 used in some places and Utf8Path in others?

File: devtools/x-lint/src/file_path.rs [hunk 3]
```diff
@@ -36,14 +37,14 @@ impl<'l> FilePathContext<'l> {
...
-    pub fn file_path(&self) -> &'l Path {
+    pub fn file_path(&self) -> &'l Utf8Path {
         &self.file_path
     }
 
     /// Returns the extension of the file. Returns `None` if there's no extension or if the
     /// extension isn't valid UTF-8.
     pub fn extension(&self) -> Option<&'l str> {
-        self.file_path.extension().map(OsStr::to_str).flatten()
+        self.file_path.extension()
```
Comment (bmwill) [inline devtools/x-lint/src/file_path.rs:47]: The comment and the invariant can probably be changed now?

File: devtools/x/src/lint/guppy.rs [hunk 4]
```diff
@@ -158,15 +158,11 @@ impl PackageLinter for CrateNamesPaths {
...
-                    "workspace path contains '_' (use '-' instead)",
-                );
-            }
-        } else {
-            // Workspace path is invalid UTF-8. A different lint should catch this.
+        if workspace_path.as_str().contains('_') {
+            out.write(
+                LintLevel::Error,
+                "workspace path contains '_' (use '-' instead)",
+            );
```
Comment (bmwill) [inline devtools/x/src/lint/guppy.rs:165]: nice!

**Review by bmwill — APPROVED** [review]
Just one clarifying question but this looks good!

File: devtools/x-lint/src/file_path.rs (hunk 3 above)
Comment (sunshowers) [reply devtools/x-lint/src/file_path.rs:47]: Ooh, will fix.

File: devtools/x-core/src/git.rs (hunk 2 above)
Comment (sunshowers) [reply devtools/x-core/src/git.rs:5]: `Utf8Paths0` stands for a list of null-separated UTF-8 paths, such as produced by `git ls-files -z`. `Utf8Path` is a single path.

File: devtools/x-lint/src/file_path.rs (hunk 3 above)
Comment (sunshowers) [reply devtools/x-lint/src/file_path.rs]: Fixed.

**Pull request merged**